File: Samples/AppServices/vb/RandomNumberService/RandomNumberGeneratorTask.vb
Imports System
Imports Windows.ApplicationModel.AppService
Imports Windows.ApplicationModel.Background
Imports Windows.Foundation.Collections

Namespace Global.RandomNumberService

    Public NotInheritable Class RandomNumberGeneratorTask
        Implements IBackgroundTask

        Dim serviceDeferral As BackgroundTaskDeferral

        Dim connection As AppServiceConnection

        Dim randomNumberGenerator As Random

        Public Sub Run(taskInstance As IBackgroundTaskInstance) Implements IBackgroundTask.Run
            serviceDeferral = taskInstance.GetDeferral()
            AddHandler taskInstance.Canceled, AddressOf OnTaskCanceled
            randomNumberGenerator = New Random(CType(DateTime.Now.Ticks, Integer))
            Dim details = TryCast(taskInstance.TriggerDetails, AppServiceTriggerDetails)
            connection = details.AppServiceConnection
            AddHandler connection.RequestReceived, AddressOf OnRequestReceived
        End Sub

        Private Sub OnTaskCanceled(sender As IBackgroundTaskInstance, reason As BackgroundTaskCancellationReason)
            If serviceDeferral IsNot Nothing Then
                serviceDeferral.Complete()
                serviceDeferral = Nothing
            End If
        End Sub

        Async Sub OnRequestReceived(sender As AppServiceConnection, args As AppServiceRequestReceivedEventArgs)
            'Get a deferral so we can use an awaitable API to respond to the message
            Dim messageDeferral = args.GetDeferral()
            Try
                Dim input = args.Request.Message
                Dim minValue As Integer = CType(input("minvalue"), Integer)
                Dim maxValue As Integer = CType(input("maxvalue"), Integer)
                'Create the response
                Dim result = New ValueSet()
                result.Add("result", randomNumberGenerator.Next(minValue, maxValue))
                Await args.Request.SendResponseAsync(result)
            Finally
                messageDeferral.Complete()
            End Try
        End Sub
    End Class
End Namespace

File: Samples/AppServices/cppwinrt/RandomNumberService/RandomNumberGeneratorTask.cpp
#include "pch.h"
#include "RandomNumberGeneratorTask.h"
#include "RandomNumberGeneratorTask.g.cpp"

using namespace winrt;
using namespace winrt::Windows::ApplicationModel::AppService;
using namespace winrt::Windows::ApplicationModel::Background;
using namespace winrt::Windows::Foundation::Collections;

namespace winrt::RandomNumberService::implementation
{
    void RandomNumberGeneratorTask::Run(IBackgroundTaskInstance const& taskInstance)
    {
        //Take a service deferral so the service isn't terminated
        serviceDeferral = taskInstance.GetDeferral();

        taskInstance.Canceled({ get_weak(), &RandomNumberGeneratorTask::OnTaskCanceled });

        //Initialize the random number generator
        randomNumberGenerator.seed(std::random_device()());

        auto details = taskInstance.TriggerDetails().as<AppServiceTriggerDetails>();
        connection = details.AppServiceConnection();

        //Listen for incoming app service requests
        connection.RequestReceived({ get_weak(), &RandomNumberGeneratorTask::OnRequestReceived });
    }

    void RandomNumberGeneratorTask::OnTaskCanceled(IBackgroundTaskInstance const&, BackgroundTaskCancellationReason const&)
    {
        if (connection != nullptr)
        {
            connection.Close();
            connection = nullptr;
        }

        if (serviceDeferral != nullptr)
        {
            //Complete the service deferral
            serviceDeferral.Complete();
            serviceDeferral = nullptr;
        }
    }

    fire_and_forget RandomNumberGeneratorTask::OnRequestReceived(AppServiceConnection const&, AppServiceRequestReceivedEventArgs const& args)
    {
        //Get a deferral so we can use an awaitable API to respond to the message
        auto messageDeferral = args.GetDeferral();

        ValueSet input = args.Request().Message();
        int minValue = input.TryLookup(L"minvalue").try_as<int>().value_or(0);
        int maxValue = input.TryLookup(L"maxvalue").try_as<int>().value_or(0);
        if (maxValue < minValue)
        {
            maxValue = minValue;
        }

        std::uniform_int_distribution<std::mt19937::result_type> distribution(minValue, maxValue); // distribution in range [minValue, maxValue]   
        int randomValue = distribution(randomNumberGenerator);

        //Create the response
        ValueSet result;
        result.Insert(L"result", box_value(randomValue));

        //Send the response
        co_await args.Request().SendResponseAsync(result);

        //Destruction of the message deferral lets the platform know we're done responding
    }

}
